<a href="https://colab.research.google.com/github/BrettonSteiner/cse380-notebooks/blob/master/10_3_About_Patterns_and_Probabilities.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Patterns and Probabilities
## Class Directed Learning
### Due: Tuesday, 9 March 2021, 11:59 pm

## TODO Explore and Wonder

Regarding spanning trees of ladder graphs:

What is the probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung?

| n | P(n) |
|---|------|
| 1 |      |
| 2 |      |
| 3 |      |
| 4 |      |
| 5 |      |


### Recurrence Relations

Let $f(n) =$ NSTIBR$(n)$:

$f(n) = 4f(n-1) - f(n-2)$ for $n > 1$;

$f(0) = 1$,

$f(1) = 1$.

Let $g(n) =$ NST$(n)$:

$g(n) = 4g(n-1) - g(n-2)$ for $n > 1$;

$g(0) = 0$,

$g(1) = 1$.

#### TODO Check Recurrences

Check the calculations in this table (maybe write recursive functions) to verify they are accurate.

Do they match what you found yesterday in your DPC?

| n | f(n) | f(n-1) | 4f(n-1)  | f(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    1 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      1 |        4 |      N/A |  N/A |
| 2 |    3 |      1 |        4 |        1 |    3 |
| 3 |   11 |      3 |       12 |        1 |   11 |
| 4 |   41 |     11 |       44 |        3 |   41 |
| 5 |  153 |     41 |      164 |       11 |  153 |

| n | g(n) | g(n-1) | 4g(n-1)  | g(n - 2) | diff |
|---|-----:|-------:|---------:|---------:|-----:|
| 0 |    0 |    N/A |      N/A |      N/A |  N/A |
| 1 |    1 |      0 |        0 |      N/A |  N/A |
| 2 |    4 |      1 |        4 |        0 |    4 |
| 3 |   15 |      4 |       16 |        1 |   15 |
| 4 |   56 |     15 |       60 |        4 |   56 |
| 5 |  209 |     56 |      224 |       15 |  209 |

In [22]:
def NST(n):
  if n == 1:
    return 1
  elif n <= 0:
    return 0
  else:
    return 4 * NST(n-1) - NST(n-2)

def NSTIBR(n):
  if n <= 1:
    return 1
  else:
    return 4 * NSTIBR(n-1) - NSTIBR(n-2)

for n in range(6):
  print(n, NSTIBR(n), NSTIBR(n - 1), 4 * NSTIBR(n - 1), NSTIBR(n - 2), 4 * NSTIBR(n - 1) - NSTIBR(n - 2))

print()
for n in range(6):
  print(n, NST(n), NST(n - 1), 4 * NST(n - 1), NST(n - 2), 4 * NST(n - 1) - NST(n - 2))

0 1 1 4 1 3
1 1 1 4 1 3
2 3 1 4 1 3
3 11 3 12 1 11
4 41 11 44 3 41
5 153 41 164 11 153

0 0 0 0 0 0
1 1 0 0 0 0
2 4 1 4 0 4
3 15 4 16 1 15
4 56 15 60 4 56
5 209 56 224 15 209


### TODO Find closed-form formulas

Can you find closed-form formulas for $f(n)$ and $g(n)$?

A closed-form formula expressing these functions in terms of operations on $n$, **without** referring to previous calculated values of the functions.

In [9]:
import math

def closed_NST(n):
  return int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n)

def closed_NSTIBR(n):
  return int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** n - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** n) - int(1/ (2 * math.sqrt(3)) * (2 + math.sqrt(3)) ** (n-1) - 1/ (2 * math.sqrt(3)) * (2 - math.sqrt(3)) ** (n-1))

print((closed_NST(1), closed_NSTIBR(1)))
print((closed_NST(2), closed_NSTIBR(2)))
print((closed_NST(3), closed_NSTIBR(3)))
print((closed_NST(4), closed_NSTIBR(4)))
print((closed_NST(5), closed_NSTIBR(5)))
print((closed_NST(6), closed_NSTIBR(6)))

(1, 1)
(4, 3)
(15, 11)
(56, 41)
(209, 153)
(780, 571)


#### Hint:

http://www.ist.tugraz.at/aichholzer/teaching/eca/spanning_trees_in_ladders.pdf

#### Possibly Illuminating Calculations

Remember continued fractions?

In [6]:
from fractions import Fraction as frac

def contfrac2frac(seq):
    """Convert the simple continued fraction in `seq`
       into a fraction with numerator num and denominator den.
    """
    num, den = 1, 0
    for u in reversed(seq):
        num, den = den + num * u, num
    return frac(num, den)

def frac2contfrac(f):
    """Build the simple continued fraction expansion of fraction f.
    """
    seq = []
    frac2contfrac_rec(f, seq)
    return seq

def frac2contfrac_rec(f, seq):
    n = f.numerator
    d = f.denominator
    if d != 0:
        seq.append(n // d)
        if n % d != 0:
            frac2contfrac_rec(frac(d, n % d), seq)

def eval_frac(f):
    """Evaluate the fraction f as a float.
    """
    return f.numerator / f.denominator

In [7]:
from math import sqrt

value = sqrt(3) - 1
value_as_cf = frac2contfrac(frac.from_float(value))[:21]
cf_to_value = contfrac2frac(value_as_cf)
print(value, value_as_cf, cf_to_value)
print(eval_frac(cf_to_value))

0.7320508075688772 [0, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2, 1, 2] 302632/413403
0.732050807565499


In [8]:
from pprint import pprint

pprint(list(map(lambda n: contfrac2frac(value_as_cf[:n]), range(20, 5, -1))))

[Fraction(110771, 151316),
 Fraction(81090, 110771),
 Fraction(29681, 40545),
 Fraction(21728, 29681),
 Fraction(7953, 10864),
 Fraction(5822, 7953),
 Fraction(2131, 2911),
 Fraction(1560, 2131),
 Fraction(571, 780),
 Fraction(418, 571),
 Fraction(153, 209),
 Fraction(112, 153),
 Fraction(41, 56),
 Fraction(30, 41),
 Fraction(11, 15)]


### Answer

In [18]:
for i in range(1, 10):  
  print(eval_frac(frac(closed_NSTIBR(i), closed_NST(i))))

print()
print(eval_frac(frac(closed_NSTIBR(100), closed_NST(100))))
print()
print(eval_frac(frac(closed_NSTIBR(200), closed_NST(200))))
print()
print(eval_frac(frac(closed_NSTIBR(500), closed_NST(500))))

1.0
0.75
0.7333333333333333
0.7321428571428571
0.7320574162679426
0.732051282051282
0.7320508416351769
0.7320261437908496
0.7320754716981132

0.7320508075688773

0.7320508075688773

0.7320508075688773


The probability ($\lim_{n \rightarrow \infty} P(n)$) that a randomly-selected spanning tree of an $n$-rung ladder graph contains the bottom rung is about 73.2%.

I worked with Jake Leung and Daniel Strickland.